### ISSUE TITLE
Token refreshing without Heroku

### ISSUE THREAD

**Issue by kaleidografik** [stevenschobert/instafeed.js#654]
Hey guys, the V2 tutorial only gives us an option to refresh our tokens via the Heroku app. However, what about if we're not able to use Heroku. Is there an alternative? I'd really like to get an automated token refreshing process in place so as to ensure continued functionality.

**Comment by stevenschobert** [comment]
Thanks for the question!

The Heroku option is currently the only open-source option we have.

But I’d really love to see the community create alternative methods in various languages and open-source them so we could think them in the README. It’d be great to see supported options in Wordpress, .NET, etc.

**Comment by kaleidografik** [comment]
Yes that would be super useful, at the moment I can't seem to find any information on how to refresh apart from via the command line, which is fine for the odd occasion but not very useful for a production environment.

**Comment by stevenschobert** [comment]
The process of getting and refreshing the access tokens is outlined in the Instagram Basic Display API docs: https://developers.facebook.com/docs/instagram-basic-display-api/guides/long-lived-access-tokens/

I think there is an example in there on using `curl` from the command line.

**Comment by benjamin-hull** [comment]
In general, a token refresh service would need to do the following things (this is assuming running on a server somewhere):

- Accept and store a valid API token as a starting point
- Periodically make a request to the IG `refresh_token` endpoint and update the current token value (every < 60 days)
- Provide some way of serving the current token value to the page where instafeed is being run. 

If this is running on the same server that is running the site you want to use instafeed on, then the simplest thing is to just include the token value in the page rendered by the server - that way there's no extra request to fetch the token, etc.

The Heroku option is really optimised for "works for most people, is free, requires next-to-no coding". It's by no means the best or most efficient way if you can do some server-side work!

**Comment by benjamin-hull** [comment]
I should also mention that we've been in contact with the developer of instant-tokens.com, which is another service designed to work with instafeed.js. As far as I know it does do everything needed, but it's centralised and not currently open-source, so I can't vouch for its security or suitability for a production environment. We'll update the readme if it that changes.

**Comment by kaleidografik** [comment]
Hey guys, just wanted to follow up on this with our implementation, which may or may not be useful for others.

So, we register our account over at instant-tokens.com, which does the heavy lifting of refreshing the short lived token every 60 days. We then use a AJAX call to hit the supplied instant-tokens.com url (which returns the access token as JSON). We can then parse the JSON to target exactly what we need, in this case `Token`.

```
$.ajax({
    type: 'get',
    dataType: 'json',
    url: 'https://ig.instant-tokens.com/users/XXXXX/instagram/XXXXX/token?userSecret=XXXXX',

    success: function (response) {
        var feed = new Instafeed({
            target: 'carousel-instagram',
            accessToken: response.Token, // Access token from json response
    });
    feed.run();
});
```

**Comment by tyssen** [comment]
It's worth pointing out that as of a few days ago Heroku is now suspending instagram-token-agent apps immediately after they're created.

https://github.com/companionstudio/instagram-token-agent/issues/13

**Comment by JonathanHo** [comment]
> Hey guys, just wanted to follow up on this with our implementation, which may or may not be useful for others.
> 
> So, we register our account over at instant-tokens.com, which does the heavy lifting of refreshing the short lived token every 60 days. We then use a AJAX call to hit the supplied instant-tokens.com url (which returns the access token as JSON). We can then parse the JSON to target exactly what we need, in this case `Token`.
> 
> ```
> $.ajax({
>     type: 'get',
>     dataType: 'json',
>     url: 'https://ig.instant-tokens.com/users/XXXXX/instagram/XXXXX/token?userSecret=XXXXX',
> 
>     success: function (response) {
>         var feed = new Instafeed({
>             target: 'carousel-instagram',
>             accessToken: response.Token, // Access token from json response
>     });
>     feed.run();
> });
> ```

Today I was getting suspended by Heroku immediately after deployment, so I switched to this method,
it worked quickly and well for me, thank you!

**Comment by idv8** [comment]
> > Hey guys, just wanted to follow up on this with our implementation, which may or may not be useful for others.
> > So, we register our account over at instant-tokens.com, which does the heavy lifting of refreshing the short lived token every 60 days. We then use a AJAX call to hit the supplied instant-tokens.com url (which returns the access token as JSON). We can then parse the JSON to target exactly what we need, in this case `Token`.
> > ```
> > $.ajax({
> >     type: 'get',
> >     dataType: 'json',
> >     url: 'https://ig.instant-tokens.com/users/XXXXX/instagram/XXXXX/token?userSecret=XXXXX',
> > 
> >     success: function (response) {
> >         var feed = new Instafeed({
> >             target: 'carousel-instagram',
> >             accessToken: response.Token, // Access token from json response
> >     });
> >     feed.run();
> > });
> > ```
> 
> Today I was getting suspended by Heroku immediately after deployment, so I switched to this method,
> it worked quickly and well for me, thank you!

Hi! I could really use a quick hand from the experts here! I want to implement instafeed.js to my personal website and am discovering that it's not as simple as it once was!
I've set up a developer account at Facebook and for Instagram. I have an account at instant-tokens.com and have my instant token URL.

The Ajax code shared above is helpful but I've had to try to convert it to vanilla javascript. I do so as a JS beginner, so I'm getting an error from instafeed.min.js:
"Uncaught (in promise) Error: accessToken must be a string or function, got null (object)"

Can you take a look and tell me where I've gone wrong? I'm sure the structure of my code is wrong.

Thank you!
Danielle

<div id="instafeed"></div>
<script type="text/javascript" src="_js/instafeed.min.js"></script>

<script type="text/javascript">
    fetch("https://ig.instant-tokens.com/users/xxx/instagram/xxxxxxxxxxxxxxxxx")
    .then(response => {
        var feed = new Instafeed({
            target: 'instafeed',
            accessToken: response.Token, // Access token from json response
        });
        feed.run();
    });
</script>

**Comment by t1mwillis** [comment]
Or, if you don't have jQuery on your site/don't want to add it, just do it using `fetch`! It is 2020 after all 😄 

```
fetch('https://ig.instant-tokens.com/users/XXXXXX/instagram/XXXXXX/token?userSecret=XXXXX')
  .then(resp => resp.json())
  .then(data => {
    const feed = new Instafeed({
            accessToken: data.Token, // Access token from json response
    });
    feed.run();
})

**Comment by garethmorgans** [comment]
Thanks @t1mwillis & @kaleidografik

**Comment by codingbadger** [comment]
Hello,

You can now link directly to a JS file using https://Instant-Tokens.com which automatically sets a JS variable named `InstagramToken` with your Instagram Access Token.

`<script src="https://ig.instant-tokens.com/users/XXXXXX/instagram/XXXXXX/token.js?userSecret=XXXXX"></script>`

https://github.com/codingbadger/instant-tokens.com/issues/2

Cheers

Barry

**Comment by Matra-Simca** [comment]
Just very easily used Instant-Tokens.com - Unlike the Heroku method I wasn't required to go through Facebook developers, does that mean this method isn't limited to the 200 user per hour restrictions?

**Comment by codingbadger** [comment]
@Matra-Simca Instant-Tokens.com uses its own Facebook application so is still bound by the same rate-limiting rules however the limit is not a per-user limit it is calculated as 240 times the number of users. 

Glad you had a good experience with instant-tokens.com :-)

**Comment by Matra-Simca** [comment]
Thanks @codingbadger - it's definitely a straightforward solution - sorry I'm still not clear in practial terms of how the rate-limiting works. Who is the user in this case, is it the end user viewing the feed, or the Instagram account linked to instant-tokens.com? Or have I misundertood completely!

Ideally I want to add two different configurations of Instafeed on different pages, it seems like this would require the same token being used in each case, so I'm wary of doubling the use rate if the practical limit is relatively low.

**Comment by codingbadger** [comment]
Instant-tokens.com is well below any rate-limiting being imposed by Instagram. As an example over 125,000 calls were made to the Instagram API via the Intstant-Tokens.com Application and it has really only scratched the surface of any limits.

**Comment by mikelivingston04** [comment]
Thanks @codingbadger - Instant-Tokens definitely seems to be the better solution. However, I cannot get it to work in my actual application. It works fine and appears to be working correctly for me within a codepen, but I get the error `Uncaught Error: Error from Instagram: The access_token provided is invalid.` in the console. The token is populating correctly, and again, all looks to be working correctly in Codepen, but i'm not sure why it wouldn't work within my own website. Any thoughts?

**Comment by codingbadger** [comment]
Can you raise an issue in https://github.com/codingbadger/instant-tokens.com/issues and provide the sample code.

Thanks
________________________________

**Comment by mikelivingston04** [comment]
@codingbadger 

Here is the issue: https://github.com/codingbadger/instant-tokens.com/issues/6. 

Not sure why the coding formatting didn't work for my code. My apologies.

**Comment by sudip-modi** [comment]
Thank you so much everyone. Finally integrated Instagram on my site🙏.

**Comment by FreMun** [comment]
A few days before I would release a website for a client, this happened: [Facebook has disabled the Instant Tokens Application](https://github.com/codingbadger/instant-tokens.com/issues/14)

Any alternative would appreciated!

**Comment by kaleidografik** [comment]
> A few days before I would release a website for a client, this happened: [Facebook has disabled the Instant Tokens Application](https://github.com/codingbadger/instant-tokens.com/issues/14)
> 
> Any alternative would appreciated!

This is such a shame, hopefully @codingbadger can get the service back up and running shortly.

**Comment by codingbadger** [comment]
The Instant Tokens Application has been reinstated by Facebook after an appeal.

**Comment by abdullahoguk** [comment]
I made a simple node.js script that refresh set of tokens. You can run the script with github actions to refresh them regularly. 
 This is an example repo that refresh in 3 days interval. Interval can be changed in workflow file (.github/workflows/tokenRefresh.yml). 
 https://github.com/abdullahoguk/refresh-instagram-tokens

**Comment by kris-g** [comment]
> I made a simple node.js script that refresh set of tokens. You can run the script with github actions to refresh them regularly. This is an example repo that refresh in 3 days interval. Interval can be changed in workflow file (.github/workflows/tokenRefresh.yml). and token list can be specified in tokens.js.
> 
> https://github.com/abdullahoguk/refresh-instagram-tokens

Is this working? I can only see the log file date being updated every 3 days. The tokens.js hasn't been updated since the initial commit.

**Comment by abdullahoguk** [comment]
Yes It's working. The tokens itself is not changing. tokens.js is a list of tokens to be reissued for batch refreshing purpose. When I debug the token, it says issued on Wednesday (3 days ago). I updated the readme for usage.

**Comment by abdullahoguk** [comment]
@kris-g 
Yes It's working. The tokens itself is not changing. tokens.js is a list of tokens to be reissued for batch refreshing purpose. When I debug the token, it says issued on Wednesday (3 days ago). I updated the readme for usage.

**Comment by jasper-clarke** [comment]
@abdullahoguk 
With your GitHub Workflow, how do you access the refreshed tokens on a website?
Do you have to manually re-copy the code from the logs every time you need it?

**Comment by abdullahoguk** [comment]
I was manually putting the token to website once only. The API was not generating a new token, it was  only extending the validity of an existing token sot hat I can use it over and over. I think that API removed in the recent API changes.

**Comment by jasper-clarke** [comment]
Thanks for getting back to me, that is really interesting.
One person I have been working with who uses your repo also said the same but they say that it is still working for them which is very strange.
We are looking into it more at the moment.

**Comment by abdullahoguk** [comment]
I have not had time to review and migrate to new API. I am getting  this error. `   "error": {
      "message": "Unsupported request - method type: get",` error in my logs with using `https://graph.instagram.com/refresh_access_token?`.
 I tried with POST but got the same error. 

 But same API endpoint successfully giving me invalidation errors  for invalidated tokens . ` "error": {
      "message": "Error validating access token: The session has been invalidated because the user changed their password or Facebook has changed the session for security reasons."`
   }

**Comment by jasper-clarke** [comment]
@abdullahoguk 
Only GET requests work for the refresh_access_token API.
That error could mean that the access token has timed out and can't be refreshed anymore or just what it says.

**Comment by philyboysmith** [comment]
Since this thread still gets traffic and the last few comments show the confusion: the ground shifted in Dec 2024 when the Basic Display API was shut down. The current state of play:

- Tokens now come from the **Instagram API with Instagram Login** flow (requires a professional/creator account — free to switch in the app settings)
- Long-lived tokens last 60 days, and `GET https://graph.instagram.com/refresh_access_token?grant_type=ig_refresh_token&access_token=...` extends them (GET only, and the token must be at least 24h old)
- If you're seeing `Unsupported request - method type: get` on that endpoint, it's almost always because the token you're passing is an old Basic Display token — those can't be refreshed anymore; you need to re-authorize under the new flow once and refresh from there

So the DIY answer to the original question ("refresh without Heroku") is: any small always-on box + a scheduled GET to the refresh endpoint. Since the refresh *extends* the same token rather than issuing a new one, you don't even need to persist anything — a monthly cron with your token is enough. I wrote up the full current flow (getting the initial token through to the refresh cron) here: https://feedframer.com/guides/instagram-long-lived-access-token

Disclosure: I built Feedframer, a hosted service that does the OAuth + refresh + caching and serves the feed as JSON/RSS (free tier, 1 account) — an option if you'd rather not babysit tokens at all. The guide above is the no-signup DIY route.
